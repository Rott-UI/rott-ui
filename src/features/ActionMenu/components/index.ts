export * from './ActionMenuComponent'
export * from './ActionMenuHeaderComponent'
