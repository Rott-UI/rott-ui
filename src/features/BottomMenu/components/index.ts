export * from './BottomMenu'
export * from './BottomMenuItem'
