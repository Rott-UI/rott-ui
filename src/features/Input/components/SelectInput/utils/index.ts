export * from './listHeightNormalizer'
export * from './modalHeightPercentageNormalizer'
