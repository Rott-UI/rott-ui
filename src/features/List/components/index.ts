export * from './List'
export * from './FavoriteSwipeComponent'
