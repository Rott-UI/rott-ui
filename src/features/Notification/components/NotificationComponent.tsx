// React Imports
import {FC} from 'react'

// Component Imports
import {Item} from '../../Item'
import {Label} from '../../Label'
import {Icon} from '../../Icon'
import {Pressable} from '../../Pressable'

// Notification Imports
import {NotificationStyle} from '../styles'

// Constant Imports
import {COLOURS} from '../../../constants'

// Package Imports
import LinearGradient from 'react-native-linear-gradient'

interface NotificationMessageProps {
  title?: string
  description?: string | JSX.Element
  iconElement: React.ReactElement | React.ReactNode
  onClose?: (() => void) | null | undefined
  variantColor: string
}

export const NotificationComponent: FC<NotificationMessageProps> = ({
  title,
  description,
  iconElement,
  variantColor,
  onClose,
}) => (
  <LinearGradient
    colors={[variantColor, COLOURS.GREY900]}
    style={NotificationStyle({variantColor}).linearGradient}
    start={{x: 0, y: 1}}
    end={{x: 0, y: 1}}
    useAngle
    angle={90}>
    <Item
      row
      relative
      width={343}
      paddingHorizontal={16}
      paddingVertical={16}
      marginBottom={4}
      marginTop={4}
      alignItemsCenter>
      <Item
        width={40}
        height={40}
        backgroundColor={COLOURS.NEUTRAL_ALPHA200}
        justifyContentCenter
        alignItemsCenter
        borderRadius={12}>
        {iconElement}
      </Item>

      <Item marginLeft={16} width={255}>
        {title && (
          <Label
            fontSize='xl'
            fontWeight={700}
            testID='notification-title-test-id'
            variant='grey-100'
            marginBottom={4}
            lineBreakMode='tail'>
            {title}
          </Label>
        )}

        {description && (
          <Label
            fontSize='sm'
            testID='notification-desc-test-id'
            variant='grey-100'
            fontWeight={500}>
            {description}
          </Label>
        )}
      </Item>

      <Pressable onPress={onClose} testID='close-icon-test-id' absolute right={10} top={10}>
        <Icon name='REMOVE' width={24} height={24} variant='grey-200' />
      </Pressable>
    </Item>
  </LinearGradient>
)
