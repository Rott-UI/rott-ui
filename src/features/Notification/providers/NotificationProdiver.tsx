// React Imports
import React, {FC, PropsWithChildren, useCallback, useMemo, useState} from 'react'

// Notification Imports
import {NotificationModel} from '../models'
import {NotificationContext} from '../contexts'
import {ToasterNotificationProvider} from './ToasterNotificationProvider'
import {useNotification} from '../hooks'
import {notificationRef} from '..'

// Package Imports
import {Toast} from 'react-native-toast-notifications'
import {NotificationStyle} from '../styles'
import {Item} from '../../Item'

let notifications = 0
export const NotificationProvider: FC<PropsWithChildren> = ({children}) => {
  const [hasAnyNotification, setHasAnyNotification] = useState<boolean>(false)

  const baseNotification = ({title, description, variant}: NotificationModel) =>
    Toast.show(title, {
      type: variant,
      data: {title, description, variant},
      onClose: () => decrement(),
    })

  const show = useCallback(
    (notification: NotificationModel) => {
      increment()

      return baseNotification(notification)
    },
    [notifications]
  )
  const hide = useCallback(
    (id: string) => {
      decrement()

      return Toast.hide(id)
    },
    [notifications]
  )
  const hideAll = useCallback(() => {
    return Toast.hideAll()
  }, [])

  const success = useCallback(
    (title: string, description?: string) => {
      increment()

      return baseNotification({title, description, variant: 'success'})
    },
    [notifications]
  )
  const warning = useCallback(
    (title: string, description?: string) => {
      increment()

      return baseNotification({title, description, variant: 'warning'})
    },
    [notifications]
  )
  const danger = useCallback(
    (title: string, description?: string) => {
      increment()

      return baseNotification({title, description, variant: 'danger'})
    },
    [notifications]
  )
  const info = useCallback(
    (title: string, description?: string) => {
      increment()

      return baseNotification({title, description, variant: 'info'})
    },
    [notifications]
  )

  const increment = () => {
    notifications = notifications + 1

    if (notifications > 0) setHasAnyNotification(true)
  }

  const decrement = () => {
    if (notifications > 0) notifications = notifications - 1

    if (notifications === 0) setHasAnyNotification(false)
  }

  const contextValues = useMemo(() => {
    return {
      notifications,
      show,
      hide,
      hideAll,

      success,
      warning,
      danger,
      info,
    }
  }, [notifications, show, hide, hideAll, success, warning, danger, info, increment, decrement])

  return (
    <NotificationContext.Provider value={contextValues}>
      <InitializeModalRef />
      {hasAnyNotification && (
        <Item testID='notification-blur-test-id' style={NotificationStyle().blur} />
      )}
      <ToasterNotificationProvider>{children}</ToasterNotificationProvider>
    </NotificationContext.Provider>
  )
}

const InitializeModalRef = () => {
  const notificationHook = useNotification()
  notificationRef.current = notificationHook

  return null
}
