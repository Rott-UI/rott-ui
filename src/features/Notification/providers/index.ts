export * from './ToasterNotificationProvider'
export * from './NotificationProdiver'
