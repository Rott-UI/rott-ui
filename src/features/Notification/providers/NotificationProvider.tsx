// React Imports
import {FC, PropsWithChildren} from 'react'

// Component Impots
import {NotificationComponent} from '../components'

// Package Imports
import {ToastProvider} from 'react-native-toast-notifications'
import {Icon} from '../../Icon'

interface NotificationProviderProps extends PropsWithChildren {}

export const NotificationProvider: FC<NotificationProviderProps> = ({children}) => {
  return (
    <ToastProvider
      placement='top'
      duration={3000}
      animationType='slide-in'
      animationDuration={250}
      offset={50}
      offsetTop={30}
      offsetBottom={40}
      swipeEnabled={true}
      renderType={{
        success: ({onHide, data: {title, description}}) => (
          <NotificationComponent
            title={title}
            description={description}
            variantColor='rgba(68, 238, 162, 0.60)'
            iconElement={
              <Icon name='CHECK_CIRCLE' width={24} height={24} mode='stroke' strokeWidth={1.5} />
            }
            onClose={() => {
              onHide()
            }}
          />
        ),
        danger: ({onHide, data: {title, description}}) => (
          <NotificationComponent
            title={title}
            description={description}
            variantColor='rgba(228, 65, 96, 0.60)'
            iconElement={
              <Icon name='WARNING_ERROR' width={24} height={24} mode='stroke' strokeWidth={1.5} />
            }
            onClose={() => {
              onHide()
            }}
          />
        ),
        warning: ({onHide, data: {title, description}}) => (
          <NotificationComponent
            title={title}
            description={description}
            variantColor='rgba(255, 117, 24, 0.60)'
            iconElement={
              <Icon name='WARNING' width={24} height={24} mode='stroke' strokeWidth={1.5} />
            }
            onClose={() => {
              onHide()
            }}
          />
        ),
        info: ({onHide, data: {title, description}}) => (
          <NotificationComponent
            title={title}
            description={description}
            variantColor='rgba(63, 182, 210, 0.60)'
            iconElement={
              <Icon name='NOTIFICATION' width={24} height={24} mode='fill' strokeWidth={1} />
            }
            onClose={() => {
              onHide()
            }}
          />
        ),
      }}>
      {children}
    </ToastProvider>
  )
}
