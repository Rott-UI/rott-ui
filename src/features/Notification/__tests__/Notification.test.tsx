// Util and Lib Imports
import {Notification} from '..'
import {render, fireEvent, act} from '../../../utils'

// Component Imports
import {NotificationComponent} from '../components'
import {NotificationProvider} from '../providers'

const mockProps = {
  title: 'Title',
  description: 'Description',
  mockOnClose: jest.fn(),
}

describe('Notification -> Custom Component', () => {
  const iconName = 'REMOVE'
  const testId = {
    closeIconTestId: 'close-icon-test-id',
    descriptionTestId: 'notification-desc-test-id',
    titleTestId: 'notification-title-test-id',
    blurTestId: 'notification-blur-test-id',
  }

  it('notification ilk render anında snapshot ile eşleşmeli', () => {
    const renderedNotification = render(
      <NotificationComponent variantColor={'white'} iconElement={iconName} {...mockProps} />
    )

    expect(renderedNotification).toMatchSnapshot()
  })

  it('notification açıldığında ekranda blur gözükmeli', () => {
    const {blurTestId} = testId
    const {getByTestId} = render(<NotificationProvider />)

    Notification.success('Test Title', 'Test Description')

    const blurElement = getByTestId(blurTestId)

    expect(blurElement).toBeOnTheScreen()
  })

  it('notification yokken ekranda blur gözükmemeli', () => {
    const {blurTestId} = testId
    const {queryByTestId} = render(<NotificationProvider />)

    const blurElement = queryByTestId(blurTestId)

    expect(blurElement).not.toBeOnTheScreen()
  })

  it('kapatma butonuna tıklandığında onClose fonksiyonları çağırılmalı', async () => {
    const {closeIconTestId} = testId
    const {getByTestId} = render(
      <NotificationComponent
        variantColor={'white'}
        iconElement={iconName}
        onClose={mockProps.mockOnClose}
        {...mockProps}
      />
    )

    await act(() => {
      const closeIcon = getByTestId(closeIconTestId)
      fireEvent.press(closeIcon)
    })

    expect(mockProps.mockOnClose).toHaveBeenCalled()
    expect(mockProps.mockOnClose).toHaveBeenCalledTimes(1)
  })

  it('title verilen değerler ile ekranda gözükmeli', () => {
    const {title} = mockProps
    const {getByText} = render(
      <NotificationComponent
        variantColor={'white'}
        iconElement={iconName}
        onClose={mockProps.mockOnClose}
        title={title}
      />
    )

    const titleElement = getByText(title)

    expect(titleElement).toBeOnTheScreen()
  })

  it('description verilen değerler ile ekranda gözükmeli', () => {
    const {description} = mockProps
    const {getByText} = render(
      <NotificationComponent
        variantColor={'white'}
        iconElement={iconName}
        onClose={mockProps.mockOnClose}
        description={description}
      />
    )

    const descriptionElement = getByText(description)

    expect(descriptionElement).toBeOnTheScreen()
  })

  it('description verilmediği zaman ekranda gözükmemeli', () => {
    const {descriptionTestId} = testId
    const {queryByTestId} = render(
      <NotificationComponent
        variantColor={'white'}
        iconElement={iconName}
        onClose={mockProps.mockOnClose}
      />
    )

    const descriptionElement = queryByTestId(descriptionTestId)

    expect(descriptionElement).not.toBeOnTheScreen()
  })
})
