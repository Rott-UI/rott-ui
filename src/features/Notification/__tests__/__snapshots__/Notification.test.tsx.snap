// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification -> Custom Component notification ilk render anında snapshot ile eşleşmeli 1`] = `
[
  <BVLinearGradient
    angle={90}
    colors={
      [
        4294967295,
        4280434502,
      ]
    }
    endPoint={
      {
        "x": 0,
        "y": 1,
      }
    }
    locations={null}
    startPoint={
      {
        "x": 0,
        "y": 1,
      }
    }
    style={
      {
        "backgroundColor": "rgba(34, 63, 70, 1)",
        "borderBottomEndRadius": 23,
        "borderBottomStartRadius": 23,
        "borderColor": "white",
        "borderEndEndRadius": 23,
        "borderEndStartRadius": 23,
        "borderRadius": 23,
        "borderStyle": "solid",
        "borderWidth": 2,
        "marginBottom": 31,
        "width": 660,
      }
    }
    useAngle={true}
  >
    <View
      alignItemsCenter={true}
      marginBottom={4}
      marginTop={4}
      paddingHorizontal={16}
      paddingVertical={16}
      relative={true}
      style={
        {
          "alignItems": "center",
          "backgroundColor": undefined,
          "bottom": undefined,
          "flex": undefined,
          "flexDirection": "row",
          "flexGrow": undefined,
          "gap": undefined,
          "height": undefined,
          "justifyContent": undefined,
          "left": undefined,
          "marginBottom": 8,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": 8,
          "maxHeight": undefined,
          "maxWidth": undefined,
          "minHeight": undefined,
          "minWidth": undefined,
          "overflow": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": 31,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "paddingVertical": 31,
          "position": "relative",
          "right": undefined,
          "top": undefined,
          "width": 660,
          "zIndex": undefined,
        }
      }
      width={343}
    >
      <View
        alignItemsCenter={true}
        backgroundColor="rgba(34, 63, 70, 0.15)"
        borderRadius={12}
        height={40}
        justifyContentCenter={true}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "rgba(34, 63, 70, 0.15)",
            "bottom": undefined,
            "flex": undefined,
            "flexDirection": "column",
            "flexGrow": undefined,
            "gap": undefined,
            "height": 77,
            "justifyContent": "center",
            "left": undefined,
            "marginBottom": undefined,
            "marginLeft": undefined,
            "marginRight": undefined,
            "marginTop": undefined,
            "maxHeight": 77,
            "maxWidth": undefined,
            "minHeight": undefined,
            "minWidth": undefined,
            "overflow": undefined,
            "paddingBottom": undefined,
            "paddingHorizontal": undefined,
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "paddingTop": undefined,
            "paddingVertical": undefined,
            "position": undefined,
            "right": undefined,
            "top": undefined,
            "width": 77,
            "zIndex": undefined,
          }
        }
        width={40}
      >
        REMOVE
      </View>
      <View
        marginLeft={16}
        style={
          {
            "alignItems": undefined,
            "backgroundColor": undefined,
            "bottom": undefined,
            "flex": undefined,
            "flexDirection": "column",
            "flexGrow": undefined,
            "gap": undefined,
            "height": undefined,
            "justifyContent": undefined,
            "left": undefined,
            "marginBottom": undefined,
            "marginLeft": 31,
            "marginRight": undefined,
            "marginTop": undefined,
            "maxHeight": undefined,
            "maxWidth": undefined,
            "minHeight": undefined,
            "minWidth": undefined,
            "overflow": undefined,
            "paddingBottom": undefined,
            "paddingHorizontal": undefined,
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "paddingTop": undefined,
            "paddingVertical": undefined,
            "position": undefined,
            "right": undefined,
            "top": undefined,
            "width": 491,
            "zIndex": undefined,
          }
        }
        width={255}
      >
        <Text
          lineBreakMode="tail"
          marginBottom={4}
          style={
            {
              "backgroundColor": undefined,
              "bottom": undefined,
              "color": "rgba(234, 239, 240, 1)",
              "flexGrow": undefined,
              "fontFamily": "Markpro-Bold",
              "fontSize": 18,
              "fontWeight": 700,
              "gap": undefined,
              "height": undefined,
              "left": undefined,
              "letterSpacing": undefined,
              "marginBottom": 8,
              "marginLeft": undefined,
              "marginRight": undefined,
              "marginTop": undefined,
              "maxHeight": undefined,
              "maxWidth": undefined,
              "minHeight": undefined,
              "minWidth": undefined,
              "overflow": undefined,
              "paddingBottom": undefined,
              "paddingHorizontal": undefined,
              "paddingLeft": undefined,
              "paddingRight": undefined,
              "paddingTop": undefined,
              "paddingVertical": undefined,
              "position": undefined,
              "right": undefined,
              "textAlign": "auto",
              "top": undefined,
              "width": undefined,
              "zIndex": undefined,
            }
          }
          testID="notification-title-test-id"
        >
          TEST WITH Test Title
        </Text>
        <Text
          style={
            {
              "backgroundColor": undefined,
              "bottom": undefined,
              "color": "rgba(234, 239, 240, 1)",
              "flexGrow": undefined,
              "fontFamily": "Markpro-Book",
              "fontSize": 12,
              "fontWeight": 500,
              "gap": undefined,
              "height": undefined,
              "left": undefined,
              "letterSpacing": undefined,
              "marginBottom": undefined,
              "marginLeft": undefined,
              "marginRight": undefined,
              "marginTop": undefined,
              "maxHeight": undefined,
              "maxWidth": undefined,
              "minHeight": undefined,
              "minWidth": undefined,
              "overflow": undefined,
              "paddingBottom": undefined,
              "paddingHorizontal": undefined,
              "paddingLeft": undefined,
              "paddingRight": undefined,
              "paddingTop": undefined,
              "paddingVertical": undefined,
              "position": undefined,
              "right": undefined,
              "textAlign": "auto",
              "top": undefined,
              "width": undefined,
              "zIndex": undefined,
            }
          }
          testID="notification-desc-test-id"
        >
          TEST WITH Test Description
        </Text>
      </View>
      <View
        absolute={true}
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        animated={false}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        right={10}
        style={
          {
            "alignItems": undefined,
            "backgroundColor": undefined,
            "bottom": undefined,
            "flex": 1,
            "flexDirection": "column",
            "flexGrow": undefined,
            "gap": undefined,
            "height": undefined,
            "justifyContent": undefined,
            "left": undefined,
            "marginBottom": undefined,
            "marginLeft": undefined,
            "marginRight": undefined,
            "marginTop": undefined,
            "maxHeight": undefined,
            "maxWidth": undefined,
            "minHeight": undefined,
            "minWidth": undefined,
            "overflow": undefined,
            "paddingBottom": undefined,
            "paddingHorizontal": undefined,
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "paddingTop": undefined,
            "paddingVertical": undefined,
            "position": "absolute",
            "right": 19,
            "top": 19,
            "width": undefined,
            "zIndex": undefined,
          }
        }
        testID="close-icon-test-id"
        top={10}
      >
        <View
          style={
            {
              "alignItems": undefined,
              "backgroundColor": undefined,
              "bottom": undefined,
              "flex": undefined,
              "flexDirection": "column",
              "flexGrow": undefined,
              "gap": undefined,
              "height": undefined,
              "justifyContent": undefined,
              "left": undefined,
              "marginBottom": undefined,
              "marginLeft": undefined,
              "marginRight": undefined,
              "marginTop": undefined,
              "maxHeight": undefined,
              "maxWidth": undefined,
              "minHeight": undefined,
              "minWidth": undefined,
              "overflow": undefined,
              "paddingBottom": undefined,
              "paddingHorizontal": undefined,
              "paddingLeft": undefined,
              "paddingRight": undefined,
              "paddingTop": undefined,
              "paddingVertical": undefined,
              "position": undefined,
              "right": undefined,
              "top": undefined,
              "width": undefined,
              "zIndex": undefined,
            }
          }
        >
          <SvgMock
            fill="transparent"
            height={46}
            stroke="rgba(161, 173, 175, 1)"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={1}
            width={46}
          />
        </View>
      </View>
    </View>
  </BVLinearGradient>,
  <View
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      [
        {
          "alignSelf": "center",
          "elevation": 999999,
          "flex": 0,
          "maxWidth": "100%",
          "position": "absolute",
          "zIndex": 999999,
        },
        {
          "flexDirection": "column-reverse",
          "justifyContent": "flex-start",
          "top": 30,
          "width": 750,
        },
      ]
    }
  >
    <View
      style={
        {
          "bottom": 0,
        }
      }
    >
      <RCTSafeAreaView />
    </View>
  </View>,
  <View
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      [
        {
          "alignSelf": "center",
          "elevation": 999999,
          "flex": 0,
          "maxWidth": "100%",
          "position": "absolute",
          "zIndex": 999999,
        },
        {
          "bottom": 40,
          "flexDirection": "column",
          "justifyContent": "flex-end",
          "width": 750,
        },
      ]
    }
  >
    <View
      style={
        {
          "bottom": 0,
        }
      }
    >
      <RCTSafeAreaView />
    </View>
  </View>,
]
`;
